src/main/java/org/purescript/psi/expression/ExpressionConstructorReference.kt

```diff
@@ -0,0 +1,39 @@
+package org.purescript.psi.expression
+
+import com.intellij.psi.PsiNamedElement
+import com.intellij.psi.PsiReferenceBase
+
+class ExpressionConstructorReference(expressionConstructor: PSExpressionConstructor) :
+    PsiReferenceBase<PSExpressionConstructor>(
+        expressionConstructor,
+        expressionConstructor.identifier.textRangeInParent,
+        false
+    ) {
+
+    override fun getVariants(): Array<Any> =
+        candidates.toList().toTypedArray()
+
+    override fun resolve(): PsiNamedElement? =
+        candidates.firstOrNull { it.name == element.name }
+
+    private val candidates: Sequence<PsiNamedElement>
+        get() {
+            val module = element.module ?: return emptySequence()
+            return sequence {
+                for (newTypeDeclaration in module.newTypeDeclarations) {
+                    yield(newTypeDeclaration.newTypeConstructor)
+                }
+                for (dataDeclaration in module.dataDeclarations) {
+                    yieldAll(dataDeclaration.dataConstructors.iterator())
+                }
+                for (importDeclaration in module.importDeclarations) {
+                    for (importedNewTypeDeclaration in importDeclaration.importedNewTypeDeclarations) {
+                        yield(importedNewTypeDeclaration.newTypeConstructor)
+                    }
+                    for (importedDataDeclaration in importDeclaration.importedDataDeclarations) {
+                        yieldAll(importedDataDeclaration.dataConstructors.iterator())
+                    }
+                }
+            }
+        }
```

Comment:
I experimented a bit with sequences, just to get a feel for the syntax. I don't think that this gives any meaningful performance improvements, but we can evaluate if we think the code is readable or not.

Reply:
In addition, I realized that this doesn't take into account cases where only the top level declaration is exported/imported, but I prefer to improve that later.

For example, this still resolves:
```purescript
module Foo where

import Data.Maybe (Maybe(Nothing))

f = Just 3
```

Reply:
i have never seen the yield syntax before, i was thinking something way simpler with `map`. Thanks for sharing and doing the reaserach!